lib/galaxy/tools/actions/__init__.py

```diff
@@ -193,7 +193,22 @@ def _collect_inputs(self, tool, trans, incoming, history, current_user_roles):
         # Collect any input datasets from the incoming parameters
         inp_data = self._collect_input_datasets(tool, incoming, trans, history=history, current_user_roles=current_user_roles)
 
-        return history, inp_data, inp_dataset_collections
+        # grap tags from incoming HDAs
+        preserved_tags = {}
+        for data in inp_data.values():
+            for tag in [t for t in data.tags if t.user_tname == 'name']:
```

Comment:
Looks like `data` can be `None` for optional inputs, so I guess you need to wrap this in an additional `if data:`

Reply:
Ahh - okay - thanks!